# FACTORIAL OF NUMBER

In [2]:
def FACTORIAL(x):
    if x==1:
        return 1
    return  x * FACTORIAL(x-1) # Recursive call 

In [3]:
FACTORIAL(5)

120

✨ **The Magic of Factorials!** ✨  

📢 **Did You Know?** The value **120** is the factorial of **5**.  

🧮 **Calculation:**  
\[
5! = 5 X 4 X 3 X 2 X 1 = 120
\]

💡 **How Does It Work?**  
Your recursive function smartly multiplies each number down to **1**, giving the perfect result — **120**! 🚀

This function calculates the **factorial** of a given number using **recursion**. Let's break it down step by step.

### Code Explanation
```python
def FACTORIAL(x):  # Function definition
    if x == 1:     # Base condition: Stops recursion when x equals 1
        return 1
    return x * FACTORIAL(x - 1)  # Recursive call
```

### Step-by-Step Working
Suppose we call the function with `FACTORIAL(4)`.

1. **First Call:** `FACTORIAL(4)`
   - Since `x != 1`, it returns `4 * FACTORIAL(3)`

2. **Second Call:** `FACTORIAL(3)`
   - Since `x != 1`, it returns `3 * FACTORIAL(2)`

3. **Third Call:** `FACTORIAL(2)`
   - Since `x != 1`, it returns `2 * FACTORIAL(1)`

4. **Fourth Call:** `FACTORIAL(1)`
   - Since `x == 1`, it returns **1** (Base condition met).

### Backtracking (Returning the Results)
- The result from `FACTORIAL(1)` (which is `1`) goes back to the previous step:
  
\[
FACTORIAL(2) = 2 \times 1 = 2
\]

- This result goes back to the previous step:

\[
FACTORIAL(3) = 3 \times 2 = 6
\]

- This result goes back to the initial call:

\[
FACTORIAL(4) = 4 \times 6 = 24
\]

### Final Answer
`FACTORIAL(4)` → **24**

### Key Concepts
✅ **Recursion**: The function calls itself.  
✅ **Base Condition**: Stops recursion when `x == 1`.  
✅ **Backtracking**: Combines results while returning from recursive calls.

# FIBONACCI SEQUENCE:

In [4]:
def FEBONACCI(n):
    fib=[0,1]
    for a in range(2,n):
        fib.append(fib[-1]+fib[-2])
    return fib    

In [5]:
FEBONACCI(5)

[0, 1, 1, 2, 3]



---

### 🔎 **Step-by-Step Example: `FEBONACCI(6)`**

1. **Initialization:**  
   `fib = [0, 1]`

---

2. **First Iteration (`a = 2`)**  
   ➔ `fib[-1] = 1` and `fib[-2] = 0`  
   ➔ `fib.append(1 + 0)`  
   ➔ **Result:** `fib = [0, 1, 1]`

---

3. **Second Iteration (`a = 3`)**  
   ➔ `fib[-1] = 1` and `fib[-2] = 1`  
   ➔ `fib.append(1 + 1)`  
   ➔ **Result:** `fib = [0, 1, 1, 2]`

---

4. **Third Iteration (`a = 4`)**  
   ➔ `fib[-1] = 2` and `fib[-2] = 1`  
   ➔ `fib.append(2 + 1)`  
   ➔ **Result:** `fib = [0, 1, 1, 2, 3]`

---

5. **Fourth Iteration (`a = 5`)**  
   ➔ `fib[-1] = 3` and `fib[-2] = 2`  
   ➔ `fib.append(3 + 2)`  
   ➔ **Result:** `fib = [0, 1, 1, 2, 3, 5]`

---

### ✨ **Final Output**
`FEBONACCI(6)` → 🌈 **[0, 1, 1, 2, 3, 5]**

---

### 📌 **Key Concepts**
✅ **Fibonacci Sequence** → Each number is the sum of the last two numbers.  
✅ **List Indexing** → `fib[-1]` and `fib[-2]` efficiently access the last two elements.  
✅ **Iteration** → Ensures numbers are generated until `n` terms are reached.

# CHECK IF THE NUMBER IS PRIME

In [293]:
def PRIME_DETECTOR(NUM):
    if NUM <= 1:  # 🔹 Numbers less than or equal to 1 are NOT prime
        return "It is not a prime number"
    else:
        for a in range(2, NUM):  # 🔹 Check divisibility from 2 to NUM-1
            if NUM % a == 0:     # 🔹 If divisible by any number in this range
                return "It is not a prime number"
                break            # 🔹 Break the loop (optional, since return exits)
        else:
            return "It is a prime number"  # 🔹 If no divisors are found


In [294]:
PRIME_DETECTOR(8)

'It is not a prime number'

# REVERSE STRING

In [295]:
def REVERSE_STRING(TEXT):
    return TEXT[::-1]
    

In [296]:
REVERSE_STRING("UJWAL")

'LAWJU'

# PALINDROME CHECK

In [297]:
def PALINDROME_CHECK(TEXT):
    return TEXT==TEXT[::-1]

In [298]:
PALINDROME_CHECK("NITIN")

True

# PALINDROME CHECK 2

In [299]:
def PALINDROME_CHECK2(TEXT):
    return TEXT=="".join(reversed(TEXT))

In [300]:
PALINDROME_CHECK2("NITIN")

True

# FIND MAX IN LIST

In [301]:
def MAX(LIST):
    return max(LIST) 
     

In [302]:
LIST=[1,2,3,4,5,6,7,20]
MAX(LIST)

20

# FIND MAX IN LIST 2

In [303]:
def MAX2(LIST):
    if not LIST:
        return None
    
    MAX_VALUE=LIST[0]
    for NUM in LIST:
        if NUM>MAX_VALUE:
            MAX_VALUE=NUM
    return MAX_VALUE    
            

In [304]:
MAX2(LIST)

20

# FIND MIN IN LIST

In [305]:
def MIN(LIST):
    return min(LIST)

In [306]:
MIN(LIST)

1

# FIND MIN IN LIST 2

In [307]:
def MIN2(LIST):
    if not LIST:
        return None
    
    MIN_VALUE=LIST[0]
    for num in LIST:
        if num<MIN_VALUE:
            MIN_VALUE=num
    return MIN_VALUE        

In [308]:
MIN2(LIST)

1

# SUM OF LIST ELEMENTS

In [309]:
def SUM(LIST):
    if not list:
        return None
    return sum(LIST)

In [310]:
SUM(LIST)

48

# SUM OF LIST ELEMENTS 2

In [311]:
def SUM2(LIST):
    if not list:
        return None
    sum=0
    for num in LIST:
        sum += num 
    return sum

In [312]:
SUM2(LIST)

48

# COUNT VOWELS IN A STRING:

In [313]:
def VOWELS_COUNT(TEXT):
    from collections import Counter
    VOWELS="AEIOUaeiou"
    FOUND_VOWELS_IN_TEXT=[char for char in TEXT if char in VOWELS]
    print(len(FOUND_VOWELS_IN_TEXT))
    print(Counter(FOUND_VOWELS_IN_TEXT))
    return sum(1 for char in TEXT if char in VOWELS)
    
    

In [314]:
TEXT="I AM UJWAL"
VOWELS_COUNT(TEXT)

4
Counter({'A': 2, 'I': 1, 'U': 1})


4

### 💡 **Key Point to Remember**  
➡️ **Any code written after `return` will not execute** because `return` immediately exits the function.  

✅ **Place print statements or additional logic before the `return` statement** if you want them to run.

# REMOVE DUPLICATES FROM LIST.

In [315]:
def REMOVE_DUPLICATE(LIST):
    return list(set(LIST))

In [316]:
LIST=["APPLE","MANGO","APPLE"]
REMOVE_DUPLICATE(LIST)

['APPLE', 'MANGO']

# COUNT THE WORDS IN A STRING

In [317]:
def WORD_COUNT(TEXT):
    return len(TEXT.split())

In [318]:
TEXT="I AM UJWAL AND I AM 21 YEAR OLD."
WORD_COUNT(TEXT)

9

# WORD COUNT 2

In [319]:
def WORD_COUNT2(TEXT):
    return TEXT.count(" ")+1

In [320]:
WORD_COUNT2(TEXT)

9

# WORD COUNT 3

In [321]:


def WORD_COUNT3(TEXT):
    from nltk.tokenize import word_tokenize
    import re
    import string
    TEXT = re.sub(r"\d+", "", TEXT)  # Removes digits
    TEXT = re.sub(r"\s+", " ", TEXT.strip())  # Removes extra spaces
    TEXT=TEXT.translate(str.maketrans("","",string.punctuation))
    TEXT = word_tokenize(TEXT)  # Tokenizes text
    return len(TEXT)


In [322]:
WORD_COUNT3(TEXT)

8

# LONGEST WORD IN A TEXT:

In [323]:
def LONGEST_WORD(TEXT):
    LIST=list(TEXT.split())
    longest_word=""
    for WORD in LIST:
        if len(WORD) > len(longest_word):
            longest_word=WORD
        
    return longest_word

In [324]:
LONGEST_WORD(TEXT)

'UJWAL'

# CAPITALIZE FIRST LETTER OF EACH WORD

In [325]:
def CAPITALIZE_WORD(TEXT):
    cap_TEXT=""
    for word in TEXT.split():
        cap_word=word.capitalize()
        cap_TEXT += cap_word+" "  
    return cap_TEXT    

In [326]:
CAPITALIZE_WORD(TEXT)

'I Am Ujwal And I Am 21 Year Old. '

# CHECK ANAGRAM

📜 **Logic to Check for Anagrams** 📜  

🔍 **Two strings are considered anagrams if:**  

✅ **Same Characters:** Both strings must contain identical characters.  
✅ **Same Frequency:** Each character must appear the same number of times.  
✅ **Order Doesn't Matter:** Character arrangement can differ.  

💡 **Example:**  
`Listen` 🔄 `Silent` → ✅ **Anagram**  
`Hello` 🔄 `Olelh` → ✅ **Anagram**  
`Python` 🔄 `Java` → ❌ **Not an Anagram**

### 🛠️ **Logic to Build an Anagram Function**  

1️⃣ **Clean the Strings**  
- Remove spaces.  
- Convert both strings to lowercase.  

2️⃣ **Check Length**  
- If the lengths are different → ❌ **Not an Anagram**  

3️⃣ **Count Characters**  
- Sort both strings.  
- If sorted versions match → ✅ **Anagram**  

---

### 🔎 **Example Flow**  
**Input:** `"Listen"` and `"Silent"`  

➡️ Cleaned Strings → `"listen"` and `"silent"`  
✅ Lowercasing → Ensures case differences don’t affect the result.

✅ Removing Spaces → Ensures extra spaces are ignored.
➡️ Length Check → **Same Length**  
➡️ Sorted Strings → `"eilnst"` and `"eilnst"`  
➡️ **Match Found** → ✅ **Anagram**

In [327]:
def CHECK_ANAGRAM(TEXT1,TEXT2):
    TEXT1=TEXT1.lower()
    TEXT2=TEXT2.lower()
    import re
    TEXT1=re.sub(r"\s+","",TEXT1)
    TEXT2=re.sub(r"\s+","",TEXT2)
    if len(TEXT1)==len(TEXT2):
        result= 1
    else:
        result= 0
    if result==1:
        if sorted(TEXT1)==sorted(TEXT2):
            return "they are anagram"
        else:
            return "they are not anagram"
    else :
        return "they are not anagram" 

In [328]:
TEXT1="i am ujwal"
TEXT2="i am ujwal"
CHECK_ANAGRAM(TEXT1,TEXT2)

'they are anagram'

# CHECK ANAGRAM 2

In [329]:
def CHECK_ANAGRAM2(TEXT1,TEXT2):
    TEXT1=TEXT1.lower()
    TEXT2=TEXT2.lower()
    import re
    TEXT1=re.sub(r"\s+","",TEXT1)
    TEXT2=re.sub(r"\s+","",TEXT2)
    if len(TEXT1) != len(TEXT2):
        return "THEY ARE NOT ANAGRAM"
    
    if sorted(TEXT1) != sorted(TEXT2):
        return "THEY ARE NOT ANAGRAM"
    
    else:
        return "THEY ARE ANAGRAM"
    

In [330]:
CHECK_ANAGRAM2(TEXT1,TEXT2)

'THEY ARE ANAGRAM'

# COUNT ALL CHARECTER OCCUERRENCE IN A TEXT:

In [331]:
def CHAR_OCCURRENCE_COUNT(TEXT):
    from collections import Counter
    return Counter(TEXT)
    

In [332]:
CHAR_OCCURRENCE_COUNT(TEXT)

Counter({' ': 8,
         'A': 5,
         'I': 2,
         'M': 2,
         'L': 2,
         'D': 2,
         'U': 1,
         'J': 1,
         'W': 1,
         'N': 1,
         '2': 1,
         '1': 1,
         'Y': 1,
         'E': 1,
         'R': 1,
         'O': 1,
         '.': 1})

# COUNT ALL CHARECTER OCCURRENCE IN A TEXT 2:

In [333]:
TEXT

'I AM UJWAL AND I AM 21 YEAR OLD.'

In [334]:
for char in set(TEXT):
    print(F"{char} = {TEXT.count(char)}") 

R = 1
O = 1
  = 8
L = 2
U = 1
A = 5
N = 1
D = 2
2 = 1
J = 1
1 = 1
Y = 1
W = 1
M = 2
I = 2
E = 1
. = 1


In [335]:
for char in set(TEXT):
    print(char)

R
O
 
L
U
A
N
D
2
J
1
Y
W
M
I
E
.


# COUNT OF SINGLE CHARECTER OCCURENCE IN A TEXT:

In [336]:
def SINGLE_CHAR_OCCURRENCE(TEXT, character):
    count_of_a = 0
    for char in TEXT:
        if char == character or char == character.lower() or char == character.upper():       # making this function case insensitive..
            count_of_a += 1
    print(f"{character} = {count_of_a}") 

In [337]:
SINGLE_CHAR_OCCURRENCE(TEXT,"A")

A = 5


# COUNT OF SINGLE CHARECTER OCCURENCE IN A TEXT 2:

In [338]:
def SINGLE_CHAR_OCCURRENCE2(TEXT, character):
    return TEXT.count(character) or TEXT.count(character.upper())
    

In [339]:
SINGLE_CHAR_OCCURRENCE2(TEXT,"a")

5

# STRING TO LIST OF CHARACTER FUNCTION:

In [340]:
def STRING_TO_LIST_CHAR(TEXT):
    return list(TEXT)

In [341]:
STRING_TO_LIST_CHAR(TEXT)

['I',
 ' ',
 'A',
 'M',
 ' ',
 'U',
 'J',
 'W',
 'A',
 'L',
 ' ',
 'A',
 'N',
 'D',
 ' ',
 'I',
 ' ',
 'A',
 'M',
 ' ',
 '2',
 '1',
 ' ',
 'Y',
 'E',
 'A',
 'R',
 ' ',
 'O',
 'L',
 'D',
 '.']

# STRING TO LIST OF CHARACTER 2

In [342]:
def STRING_TO_LIST_CHAR2(TEXT):
    LIST=[char for char in TEXT]
    return LIST

In [343]:
STRING_TO_LIST_CHAR2(TEXT)

['I',
 ' ',
 'A',
 'M',
 ' ',
 'U',
 'J',
 'W',
 'A',
 'L',
 ' ',
 'A',
 'N',
 'D',
 ' ',
 'I',
 ' ',
 'A',
 'M',
 ' ',
 '2',
 '1',
 ' ',
 'Y',
 'E',
 'A',
 'R',
 ' ',
 'O',
 'L',
 'D',
 '.']

# REPLACE SUBSTRING(CASE SENSITIVE)

In [344]:
def REPLACE_SUBSTRING(TEXT,OLD,NEW):
    TEXT = TEXT.replace(OLD,NEW)
    return TEXT

In [345]:
TEXT

'I AM UJWAL AND I AM 21 YEAR OLD.'

In [346]:
REPLACE_SUBSTRING(TEXT,"UJWAL","nagesh")

'I AM nagesh AND I AM 21 YEAR OLD.'

# REPLACE SUBSTRING (CASE INSENSITIVE)

In [347]:
def REPLACE_SUBSTRING_INSENSITIVE(TEXT,OLD,NEW):
    if OLD in TEXT:
        TEXT=TEXT.replace(OLD,NEW) # replace function does not modify the original text directly
        return TEXT
    elif OLD.lower() in TEXT:
        TEXT_l=TEXT.replace(OLD.lower(),NEW)
        return TEXT_l
    elif OLD.upper() in TEXT:
        TEXT_u=TEXT.replace(OLD.upper(),NEW)
        return TEXT_u            

In [348]:
REPLACE_SUBSTRING_INSENSITIVE(TEXT,"nagesh","UJWAL")

In [349]:
TEXT

'I AM UJWAL AND I AM 21 YEAR OLD.'

# REPLACE SUBSTRING (CASE INSENSITIVE)2

In [350]:
def REPLACE_SUBSTRING_INSENSITIVE2(TEXT, OLD, NEW):
    # Use re.IGNORECASE to handle all case variations
    import re
    return re.sub(re.escape(OLD), NEW, TEXT, flags=re.IGNORECASE)


In [351]:
REPLACE_SUBSTRING_INSENSITIVE2(TEXT,"UJWAL","nagesh")

'I AM nagesh AND I AM 21 YEAR OLD.'

# REPLACE SUBSTRING (CASE INSENSITIVE)3

In [352]:
def REPLACE_SUBSTRING_INSENSITIVE3(TEXT,OLD,NEW):
    text=TEXT.lower()
    old=OLD.lower()
    new=NEW.lower()
    TEXT=text.replace(old,new)
    return TEXT

In [353]:
REPLACE_SUBSTRING_INSENSITIVE3(TEXT,"nagesh","UJWAL")

'i am ujwal and i am 21 year old.'

In [354]:
TEXT="I AM UJWAL AND I AM 21 YEAR OLD."

# REPLACE SUBSTRING THE NEW WORD SHOULD BE ACCORDING TO CASE OF SENTENCE.

In [355]:
def REPLACE_SUBSTRING2(TEXT, OLD, NEW):
    def CASE_NEW(NEW):
        if OLD.isupper():
            return NEW.upper()
        elif OLD.islower():
            return NEW.lower()
        else:
            return NEW.capitalize()

    TEXT = TEXT.replace(OLD, CASE_NEW(NEW))  
    return TEXT


In [356]:
TEXT

'I AM UJWAL AND I AM 21 YEAR OLD.'

In [358]:
REPLACE_SUBSTRING2(TEXT,"UJWAL","nagesh")

'I AM NAGESH AND I AM 21 YEAR OLD.'